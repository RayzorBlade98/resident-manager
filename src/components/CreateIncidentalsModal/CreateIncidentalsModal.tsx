import React, { useState } from "react";
import GenericModal from "../GenericComponents/GenericModal/GenericModal";
import {
  DeductionType,
  emptyIncidentals,
  Incidentals,
} from "_/types/incidentals";
import { addIncidentals } from "_/states/saveStates/incidentals_state";
import {
  Button,
  Grid,
  InputAdornment,
  MenuItem,
  TextField,
} from "@mui/material";
import { convert_currency_float_to_int } from "_/utils/currency";

export interface CreateIncidentalsModalProps {
  show: boolean;
  onClose: () => void;
}

function CreateIncidentalsModal(
  props: CreateIncidentalsModalProps
): JSX.Element {
  const [incidentals, setIncidentals] = useState<Incidentals>(
    emptyIncidentals()
  );

  function incidentalsUpdater(field: string) {
    function updateIncidentals(
      event: React.ChangeEvent<HTMLInputElement>
    ): void {
      let value: number | string = event.target.value;
      if (["currentPrice", "invoiceInterval"].includes(field)) {
        value = Number(value);
        if (field === "currentPrice")
          value = convert_currency_float_to_int(value);
      }

      setIncidentals({
        ...incidentals,
        [field]: value,
      });
    }
    return updateIncidentals;
  }

  function onSave(): void {
    addIncidentals(incidentals);
    props.onClose();
  }

  return (
    <GenericModal title="Neue Nebenkosten" {...props}>
      {/*Body*/}
      <form>
        <Grid container columnSpacing={2} rowSpacing={2}>
          <Grid item xs={6}>
            <TextField
              id="name"
              label="Name"
              variant="outlined"
              required
              onChange={incidentalsUpdater("name")}
            />
          </Grid>
          <Grid item xs={6}>
            <TextField
              id="cost"
              label="Kosten"
              type="number"
              required
              onChange={incidentalsUpdater("currentPrice")}
              InputProps={{
                endAdornment: <InputAdornment position="end">â‚¬</InputAdornment>,
              }}
            />
          </Grid>
          <Grid item xs={6}>
            <TextField
              id="deductionType"
              select
              required
              label="Abrechnungsart"
              value={incidentals.deductionType}
              onChange={incidentalsUpdater("deductionType")}
            >
              {Object.values(DeductionType).map((type) => (
                <MenuItem key={type} value={type}>
                  {type}
                </MenuItem>
              ))}
            </TextField>
          </Grid>
          <Grid item xs={6}>
            <TextField
              label="Abrechnungszeitraum"
              id="invoiceInterval"
              type="number"
              required
            />
          </Grid>
        </Grid>
      </form>
      {/*Footer*/}
      <Button variant="contained" onClick={onSave}>
        Erstellen
      </Button>
    </GenericModal>
  );
}

export default CreateIncidentalsModal;
