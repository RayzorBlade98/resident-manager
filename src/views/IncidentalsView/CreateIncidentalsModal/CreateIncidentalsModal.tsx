import React from 'react';
import { useRecoilValue, useResetRecoilState } from 'recoil';
import GenericModal from '../../../components/GenericComponents/GenericModal/GenericModal';
import createIncidentalsState from '../states/create_incidentals_state';
import CreateIncidentalsButton from './CreateIncidentalsButton/CreateIncidentalsButton';
import CreateIncidentalsForm from './CreateIncidentalsForm/CreateIncidentalsForm';

/**
 * Modal that contains an input form to create new incidentals.
 */
function CreateIncidentalsModal(): JSX.Element {
  const incidentalsCreationState = useRecoilValue(createIncidentalsState);
  const resetIncidentalsCreationState = useResetRecoilState(
    createIncidentalsState,
  );

  return (
    <GenericModal
      title="Neue Nebenkosten"
      show={incidentalsCreationState.showModal}
      onClose={resetIncidentalsCreationState}
    >
      {/* Body */}
      <CreateIncidentalsForm />
      {/* Footer */}
      <CreateIncidentalsButton />
    </GenericModal>
  );
}

export default CreateIncidentalsModal;
