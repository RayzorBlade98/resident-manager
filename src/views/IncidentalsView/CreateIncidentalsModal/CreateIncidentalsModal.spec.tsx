import { act, render } from '@testing-library/react';
import React from 'react';
import { setRecoil } from 'recoil-nexus';
import createIncidentalsState from '../states/create_incidentals_state';
import CreateIncidentalsModal from './CreateIncidentalsModal';
import ReactTestWrapper from '_/test/ReactTestWrapper';

describe('CreateIncidentalsModal', () => {
  test('should match snapshot', () => {
    // Arrange
    const renderResult = render(
      <ReactTestWrapper>
        <CreateIncidentalsModal />
      </ReactTestWrapper>,
    );
    act(() => {
      setRecoil(createIncidentalsState, (state) => ({
        ...state,
        showModal: true,
      }));
    });

    // Assert
    expect(renderResult.baseElement).toMatchSnapshot();
  });
});
