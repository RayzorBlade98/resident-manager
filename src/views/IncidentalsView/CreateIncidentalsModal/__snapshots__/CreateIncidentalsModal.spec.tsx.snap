// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateIncidentalsModal should match snapshot 1`] = `
<div>
  <div
    class="modal fade show"
    style="display: block; pointer-events: initial;"
    tabindex="-1"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            Neue Nebenkosten
          </h5>
          <button
            class="ripple ripple-surface btn-close"
            role="button"
          />
        </div>
        <div
          class="modal-body"
        >
          <div
            class="MuiGrid-root MuiGrid-container css-mhc70k-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="name"
                  id="name-label"
                >
                  Name
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="name"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Name
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="cost"
                  id="cost-label"
                >
                  Kosten
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="cost"
                    required=""
                    type="number"
                    value=""
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                    >
                      €
                    </p>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Kosten
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  id="deductionTypeLabel"
                >
                  Abrechnungsart
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                  required=""
                >
                  <div
                    aria-expanded="false"
                    aria-haspopup="listbox"
                    aria-labelledby="deductionTypeLabel deductionType"
                    class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                    id="deductionType"
                    role="button"
                    tabindex="0"
                  >
                    Pro Wohnung
                  </div>
                  <input
                    aria-hidden="true"
                    class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                    required=""
                    tabindex="-1"
                    value="Pro Wohnung"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-14lo706"
                    >
                      <span>
                        Abrechnungsart
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="invoiceInterval"
                  id="invoiceInterval-label"
                >
                  Abrechnungszeitraum
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id="invoiceInterval"
                    max="12"
                    min="1"
                    required=""
                    type="number"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Abrechnungszeitraum
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-focusVisible MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Erstellen
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            >
              <span
                class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
